### PR TITLE
fix(helm): properly compare existing and candidate releases

### PR DESCRIPTION
**Pull request by cndoit18**
Signed-off-by: cndoit18 <cndoit18@outlook.com>

**Description of the change:**

Fix when using the randAlphaNum in Helm, the chart release is constantly upgraded

**Motivation for the change:**

fixes: #4936 

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))

### CODE REVIEW

File: internal/helm/release/manager.go [hunk 1]
```diff
@@ -125,15 +126,12 @@ func (m *manager) Sync(ctx context.Context) error {
...
 
-	// Get the next candidate release to determine if an upgrade is necessary.
-	candidateRelease, err := m.getCandidateRelease(m.namespace, m.releaseName, m.chart, m.values)
-	if err != nil {
-		return fmt.Errorf("failed to get candidate release: %w", err)
-	}
-	if deployedRelease.Manifest != candidateRelease.Manifest {
-		m.isUpgradeRequired = true
-	}
+	upgrade := m.namespace == deployedRelease.Namespace
```
Comment (estroz) [inline internal/helm/release/manager.go:129]: Can you add a comment and a test case for this?

**Review by estroz — CHANGES_REQUESTED** [review]
It is not exactly clear why this fixes the issue described. In this case, comparing manifests will not work because the RNG will generate two different numbers, and everything else is the same?

**Comment by cndoit18** [comment]
> It is not exactly clear why this fixes the issue described. In this case, comparing manifests will not work because the RNG will generate two different numbers, and everything else is the same?

yes. and I will add this unit test.

**Comment by cndoit18** [comment]
> It is not exactly clear why this fixes the issue described. In this case, comparing manifests will not work because the RNG will generate two different numbers, and everything else is the same?

Yes, It will cause the helm-operator always re-upgrade

File: internal/helm/release/manager.go (hunk 1 above)
Comment (cndoit18) [reply internal/helm/release/manager.go:129]: > Can you add a comment and a test case for this?

Yes. I need some time

File: changelog/fragments/helm-randalphanum.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+entries:
+  - description: >
+      Fix when using the randAlphaNum in Helm, the chart release is constantly upgraded
```
Comment (estroz) [inline changelog/fragments/helm-randalphanum.yaml:3]:
```suggestion
      For Helm-based operators, fixed a bug where deployed and candidate release comparison was always false when an RNG was used to derive some manifest value, resulting in the chart release constantly upgrading
```

**Review by estroz — CHANGES_REQUESTED** [review]
Nit on changelog text, otherwise LGTM

**Review by estroz — APPROVED** [review]
Thanks @cndoit18!
/lgtm

**Comment by cndoit18** [comment]
> Thanks @cndoit18!
> /lgtm

I resubmitted and fixed the problem with the e2e docs

**Comment by cndoit18** [comment]
> Thanks @cndoit18!
> /lgtm

Can you please perform ci first? Just like `need-test-ok` in the k8s community

**Comment by cndoit18** [comment]
> > Thanks @cndoit18!
> > /lgtm
> 
> I resubmitted and fixed the problem with the e2e docs

I'm having some problems with the ci and it doesn't look too much like I'm the cause. Can you help me? please
@estroz

**Comment by estroz** [comment]
@cndoit18 the goreportcard link is bad, nothing to do on your end. I will force merge this.

**Pull request merged**